plugins/amqp_trace_plugin/include/eosio/amqp_trace_plugin/amqp_handler.hpp

```diff
@@ -0,0 +1,95 @@
+#pragma once
+#include <amqpcpp.h>
+#include <amqpcpp/libboostasio.h>
+#include <amqpcpp/linux_tcp.h>
+
+#include <fc/log/logger.hpp>
+
+namespace eosio {
+
+/// Designed to work with io_service running on a dedicated thread.
+class amqp {
+public:
+   using on_error_t = std::function<void(const std::string& err)>;
+   amqp( boost::asio::io_service& io_service, const std::string& address, std::string name, on_error_t on_err )
+         : name_( std::move( name ) )
+         , on_error( std::move(on_err) )
+   {
+      AMQP::Address amqp_address( address );
+      ilog( "Connecting to AMQP address ${a} - Queue: ${q}...", ("a", std::string( amqp_address ))( "q", name_ ) );
+
+      handler_ = std::make_unique<amqp_handler>( *this, io_service );
+      connection_ = std::make_unique<AMQP::TcpConnection>( handler_.get(), amqp_address );
+      channel_ = std::make_unique<AMQP::TcpChannel>( connection_.get() );
```

Comment:
Is this safe to run in the application thead?  I'd guess that it isn't, but I'm not sure.

Reply:
Refactored to run in amqp thread.